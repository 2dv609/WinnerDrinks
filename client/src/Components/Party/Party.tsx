import React from 'react';
import { TextGameModuleProps } from '../GameModuleProps';


const Party: React.FC<TextGameModuleProps> = ({ gameService, gameEvent, currentPlayers }) => {

    function addPlayerToGameEvent(): string {
        return gameEvent.question.replace(/{p0}/g, `${currentPlayers[0]}`).replace(/{p1}/g, `${currentPlayers[1]}`)
    }

    return (
        <div className="card block">
            <div className="card-content">
                <div className="content" data-testid="game-event">{addPlayerToGameEvent()}</div>
                <h3>Who won?</h3>
                <div className="columns">
                    <div className="column">
                        <button className="button" data-testid="button-1" onClick={() => {
                            gameService.addScore(currentPlayers[1])
                            gameService.makeWinnerAlert(currentPlayers[1])
                            gameService.chooseRandomNewGame()
                        }}>{currentPlayers[1].toString()}</button>
                    </div>
                    <div className="column">
                        <button className="button" data-testid="button-0" onClick={() => {
                            gameService.addScore(currentPlayers[0])
                            gameService.makeWinnerAlert(currentPlayers[0])
                            gameService.chooseRandomNewGame()
                        }}>{currentPlayers[0].toString()}</button>
                    </div>
                </div>




            </div>
        </div>

    )
};

export default Party;