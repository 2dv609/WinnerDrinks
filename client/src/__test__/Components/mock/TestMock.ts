import { GameModuleProps } from '../../../Components/GameModuleProps'
import Player from '../../../model/Player'

export const playersMock: Player[] = [new Player('Tom'), new Player('Carl'), new Player('Sandra')]

const makeWinnerAlertMock = (p: Player | Player[] | null, message?: string): void => {}
const chooseRandomNewGameMock = (): void => {}
const removeGameEventMock = (gameEventId: string, gameEvents: GameEventAPI): void => {}
const getPlayersMock = (amount: number, players: Player[]): Player[] => {return players}
const addScoreMock = (p: Player) => void {}


export const gameServiceMock: GameModuleProps = {
    players: playersMock,
    getPlayers: getPlayersMock,
    addScore: addScoreMock,
    makeWinnerAlert: makeWinnerAlertMock,
    chooseRandomNewGame: chooseRandomNewGameMock,
    removeGameEvent: removeGameEventMock        
}
