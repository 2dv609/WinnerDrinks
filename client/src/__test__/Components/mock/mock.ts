import Player from '../../../Player'


// Mock GameProps
const addScoreMock = (p: Player, score: number) => {}
const makeWinnerAlertMock = (p: any) => {}
const chooseRandomNewGameMock = () => {}
const getPlayersMock = (amount: number): Player[] => {
    const player1 = new Player('Mats')
    const player2 = new Player('Tom')
    return [player1, player2];
};

export const gamePropsMock = {
    getPlayers: getPlayersMock, // Let the component decide how many players it needs. 
    addScore: addScoreMock, // Adds a score to a player
    makeWinnerAlert: makeWinnerAlertMock, // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: chooseRandomNewGameMock, // chooses a random new game
}

// Mock game event mock
export const gameEventMock = {
    _id: 111,
    question: 'This is a mock question'
}