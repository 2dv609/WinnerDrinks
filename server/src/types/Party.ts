import { Document } from 'mongoose'

export interface IParty extends Document {
  question: string
}
