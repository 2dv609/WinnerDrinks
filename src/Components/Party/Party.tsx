import React, { useState, useEffect } from 'react';
import GameProps from '../GameProps';
import '../../App.css';

function Party(props: any) {
    const gp: GameProps = props.gp;
    const [players, setPlayers] = useState(gp.getPlayers(2));
    


    const tasks: string[] = [
        `${players[0]} och ${players[1]} spelar sten-sax-påse, vinnaren tar en shot`,
        `${players[0]} och ${players[1]} dricker 10000 shots`,
        `${players[0]} kittlar ${players[1]} annars sprängs solen`,
        `${players[0]} och ${players[1]} pratar känslor`,
        `${players[0]} och ${players[1]} inser livets mening eller tar en shot`,

    ];

    const [task, setTask] = useState(getNewTask());

    function getNewTask(): string {
        const rand = Math.floor(Math.random() * tasks.length)
        return tasks[rand];
    }

    useEffect(() => { 
        return () => { // Return a function for code cleanup. This will set new players 
            setPlayers(gp.getPlayers(2));
        }
    }, [gp])

    return (
        <div>
            <div className="task">{task}</div>
            <button onClick={() => setTask(getNewTask())} >Click here to get a task</button>
            <h3>Who won?</h3>

            <button onClick={() => {
                gp.done(players[0], 1)
            }}>{players[0].toString()}</button>

            <button onClick={() => {
                gp.done(players[1], 1)
            }}>{players[1].toString()}</button>

        </div>
    )
};

export default Party;