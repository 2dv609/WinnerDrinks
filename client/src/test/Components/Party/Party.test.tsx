import React from 'react';
import ReactDOM from 'react-dom';
import Party from '../../../Components/Party/Party'
import { gamePropsMock, gameEventMock } from '../mock/mock'
import { render, cleanup } from '@testing-library/react' 
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'


// After each cleanup
afterEach(cleanup)

// Test if component do not crashes with common interface. 
it('Game module should use common interface', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Party gp={gamePropsMock} gameEvent={gameEventMock}></Party>, div)
})

// Test if componenet render correct gameEvent
it('Game module should render the correct game event', () => {
    const { getByTestId } = render(<Party gp={gamePropsMock} gameEvent={gameEventMock}></Party>)
    expect(getByTestId('game-event')).toHaveTextContent(addPlayerToGameEvent(gameEventMock))
})

// Create a snapshot of the element
it('Game module should match snapshot', () => {
    const tree = renderer.create(<Party gp={gamePropsMock} gameEvent={gameEventMock}></Party>).toJSON()
    expect(tree).toMatchSnapshot()
})

// Helper function
function addPlayerToGameEvent(gameEvent: any): string {
    const gameEventQuestion: string = gameEvent.question
    return gameEventQuestion.replace('{players[0]}', `${gamePropsMock.getPlayers(1)[0]}`).replace('{players[1]}', `${gamePropsMock.getPlayers(1)[1]}}`)
}