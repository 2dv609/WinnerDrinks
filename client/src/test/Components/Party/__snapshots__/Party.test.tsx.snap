// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game module should match snapshot 1`] = `
<div
  className="box"
>
  <div
    className="content"
    data-testid="game-event"
  >
    This is a mock question
  </div>
  <h3>
    Who won?
  </h3>
  <button
    className="button"
    onClick={[Function]}
  >
    Mats
  </button>
  <button
    className="button"
    onClick={[Function]}
  >
    Tom
  </button>
</div>
`;
